### PR TITLE
Adds support for *.sam.gz and *.sam.xz in calculate_coverage.py

### PR DESCRIPTION
**Pull request by dhakim87**
Right now this only works for calculate_coverage.py.  I think making something similar for filter_sam would be straightforward, definitely easy to add this exact change to read in files, but I didn't feel comfortable deciding whether output files should be zipped or not matching the input files.  

Let me know if you want something like this and I'm happy to help add it in.

### CODE REVIEW

File: calculate_coverages.py [hunk 1]
```diff
@@ -17,21 +21,38 @@ def calculate_coverages(input, output, database):
...
+            open_sam_file = gzip.open(
+                samfile.strip(),
+                mode='rt',
+                encoding='utf-8')
+        elif samfile.endswith(".sam.xz"):
+            open_sam_file = lzma.open(
+                samfile.strip(),
+                mode='rt',
+                encoding='utf-8')
+
```
Comment (wasade) [inline calculate_coverages.py:42]: ...shouldn't happen but:

```suggestion
else:
    raise IOError("Unrecognized file extension on '%s'." % samfile)
```

File: calculate_coverages.py [hunk 2]
```diff
@@ -17,21 +21,38 @@ def calculate_coverages(input, output, database):
...
+        with open_sam_file:
```
Comment (wasade) [inline calculate_coverages.py:43]: I don't think the context is necessary as the file was opened explicitly?

File: calculate_coverages.py [hunk 3]
```diff
@@ -17,21 +21,38 @@ def calculate_coverages(input, output, database):
...
                 gotu = linesplit[2]
                 location = int(linesplit[3])
-                #Get sum of lengths in CIGAR string. Counting deleitons as alignment because they should be small
+                #Get sum of lengths in CIGAR string. Counting deletions as alignment because they should be small
                 length_string = linesplit[5]
                 length = sum([int(x) for x in re.split("[a-zA-Z]",length_string) if x])
                 #Add range to contig_dict
-                for i in range(location,location+length):
+                for i in range(location, location+length):
                     gotu_dict[gotu].add(i)
```
Comment (wasade) [inline calculate_coverages.py:55]: I think an explicit close should be issued after use
```suggestion
        open_sam_file.close()
```

**Review by wasade — CHANGES_REQUESTED** [review]
Just some minor items

File: calculate_coverages.py (hunk 2 above)
Comment (dhakim87) [reply calculate_coverages.py:44]: My intention was to let the with statement automatically close the open_sam_file, avoiding the need for exception handling and an explicit close() call.  Does that work?

Comment (dhakim87) [reply calculate_coverages.py:44]: with statement seems to still call __enter__ and __exit__ even if target object was constructed prior:

`class Foo:
    def __enter__(self):
        print("Entering Foo")

    def __exit__(self, exc_type, exc_val, exc_tb):
        print("Exiting Foo")


def open_it():
    print("Opening Foo")
    foo = Foo()
    print("Returning Foo")
    return foo

print("CASE 1, with open_it as foo:")
with open_it() as foo:
    print("Holding onto foo")

print("CASE 2, with existing foo:")
foo = open_it()
with foo:
    print("Holding onto foo")
`

Comment (wasade) [reply calculate_coverages.py:44]: oh weird. okay

**Review by wasade — APPROVED** [review]

**Comment by wasade** [comment]
I'm not sure who is maintainer -- @adswafford, are you good if I merge items?

**Comment by dhakim87** [comment]
Should hold off on merge until there is a strategy for adding something similar to filter_sam.py?

**Comment by adswafford** [comment]
After merging other branch, conflicts with code. @dhakim87 can you have a look and resolve?

**Comment by dhakim87** [comment]
Should be resolved

**Comment by adswafford** [comment]
@dhakim87 did you get a chance to update the filter_sam.py to take sam.xz and sam.gz files as well? Relatedly, do you have your code snippet for running Woltka afterward on the filtered sams that we can add to the documentation?

**Comment by dhakim87** [comment]
I've been blocked hoping to ask maintainer about choices for output file format:

If filter_sam.py is given a .sam.xz, what should it produce as output? 
- **A:** .sam 
- **B:** .sam.xz  

Will be trivial to add once that is decided.

**Comment by dhakim87** [comment]
> @dhakim87 ... Relatedly, do you have your code snippet for running Woltka afterward on the filtered sams that we can add to the documentation?

I'm not sure what you mean by this exactly.  filter_sam.py just produces new sam files, so any tool that takes a .sam will run exactly the same as before.  I've never actually run filter_sam.py, I've only looked at the coverage results and filtered the resulting biom table.  

Do you mean the code I used to import the coverage output into my mimics database for the momi tool?  That would just be converting to sql as per:

```
import sqlite3
import csv

def add_zebra():
    conn = sqlite3.connect("mimics.db")
    c = conn.cursor()
    c.execute("CREATE TABLE IF NOT EXISTS zebra(genome_id text PRIMARY KEY, covered_length int, total_length int, coverage_ratio real)")
    with open("coverage_output.tsv") as zebra_output:
        zebra_reader = csv.reader(zebra_output, delimiter='\t', quotechar='|')
        header = next(zebra_reader)
        for row in zebra_reader:
            c.execute("INSERT INTO zebra VALUES(?,?,?,?)", (row[0], row[1], row[2], row[3]))
    conn.commit()
    conn.close()
```

**Comment by adswafford** [comment]
Ah, filtering that way isn't advisable imo since you don't get redistribution of the reads to the retained list of microbes. Since woltka takes sam.xz I think outputting in that format is desirable.

**Comment by dhakim87** [comment]
> Ah, filtering that way isn't advisable imo since you don't get redistribution of the reads to the retained list of microbes. Since woltka takes sam.xz I think outputting in that format is desirable.

Does woltka look over individual files or all files for a list of included matches at any point?  If not, I don't understand how it can redistribute reads in a meaningful way- all the reads associated with the genomes that don't match the threshold criteria are entirely removed in filter_sam aren't they?

**Comment by gregpoore** [comment]
Does woltka look over individual files or all files for a list of included matches at any point? If not, I don't understand how it can redistribute reads in a meaningful way- all the reads associated with the genomes that don't match the threshold criteria are entirely removed in filter_sam aren't they?

Yes, but in the case where Woltka splits a read among several possible assignments, the redistribution post filtering will change versus the original assignments.

On May 26, 2021, at 2:48 PM, Daniel Hakim ***@***.******@***.***>> wrote:



Ah, filtering that way isn't advisable imo since you don't get redistribution of the reads to the retained list of microbes. Since woltka takes sam.xz I think outputting in that format is desirable.

Does woltka look over individual files or all files for a list of included matches at any point? If not, I don't understand how it can redistribute reads in a meaningful way- all the reads associated with the genomes that don't match the threshold criteria are entirely removed in filter_sam aren't they?

**Comment by adswafford** [comment]
That's correct. So filtering the biom table will artificially deflate the total read count since all of the fractional assignments to zebras are discarded. If you run filter sam however and supply the output sam(.xz) to Woltka, then they will be correctly distributed. 

For a toy example, imagine 1000 reads each align to 10 different microbes. So in the biom table that read is distributed across the microbes with each one receiving 100 reads.  If you then calc coverages and eliminate 9 of those microbes from the biom table, you'll be down to 100 reads for the one remaining microbe. If you instead filter_sam.py that one microbe will now get all 1000 reads.

Does that make sense?

**Comment by dhakim87** [comment]
My gut feeling is that something is off about this high level procedure.  In the edge case that there is a read that aligns to 16 different species, none of which meet the coverage threshold, that read is silently discarded at the filter_sam layer.  Ideally that error case would bubble up through woltka by instead making woltka aware of the coverage threshold and which reference genomes pass and fail that threshold.  

That has exactly nothing to do with zip support though

**Pull request merged**